ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala

```diff
@@ -0,0 +1,166 @@
+        .map(fetchFuture(_, applicationId, parties))
+        .getOrElse(futureEmptyList)
+
+    val filteredCache = inMemCache
+      .slice(requestedRange)
+      .map(cache => filterAndMapToResponse(cache.cache.toSeq, applicationId, parties))
+      .getOrElse(Seq.empty)
+    for {
+      historicCompletions <- historicCompletionsFuture
+      futureCompletions <- futureCompletionsFuture
+    } yield historicCompletions ++ filteredCache ++ futureCompletions
+  }
+
+  /** fetches completions older than start offset of cache
+    */
+  private def fetchHistoric(
+      range: Range,
+      applicationId: ApplicationId,
+      parties: Set[Ref.Party],
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] =
+    completionsDao
+      .getFilteredCompletions(
+        range.startExclusive,
+        range.endInclusive,
+        applicationId,
+        parties,
+      )
+
+  private def calculateHistoricRangeToFetch(
+      cachedRange: Range,
+      requestedRange: Range,
+  ): Option[Range] = requestedRange.lesserRangeDifference(cachedRange)
+
+  /** fetches completions ahead cache and caches results
+    */
+  private def fetchFuture(
+      rangeToFetch: Range,
+      applicationId: ApplicationId,
+      parties: Set[Ref.Party],
+  )(implicit
+      loggingContext: LoggingContext,
+      executionContext: ExecutionContext,
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] =
+    pendingRequestRef.synchronized {
+      val pendingRequest = pendingRequestRef.get()
+      if (pendingRequest.isCompleted) {
+        val allCompletionsFuture = createAndSetNewPendingRequest(rangeToFetch)
+        allCompletionsFuture.map(filterAndMapToResponse(_, applicationId, parties))
+      } else {
+        pendingRequest.flatMap { _ =>
+          getCompletionsPage(
+            rangeToFetch.startExclusive,
+            rangeToFetch.endInclusive,
+            applicationId,
+            parties,
+          )
+        }
+      }
+    }
+
+  private def filterAndMapToResponse(
+      completions: Seq[(Offset, CompletionStreamResponseWithParties)],
+      applicationId: ApplicationId,
+      parties: Set[Ref.Party],
+  ): Seq[(Offset, CompletionStreamResponse)] =
+    completions
+      .filter { case (_, completion) =>
+        completion.applicationId == applicationId && completion.parties.exists(
+          parties.contains
+        )
+      }
+      .map { case (offset, completionWithParties) =>
+        (offset, completionWithParties.completion)
+      }
+
+  private def createAndSetNewPendingRequest(
+      requestedRange: Range
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[List[(Offset, CompletionStreamResponseWithParties)]] = {
+    val allCompletionsFuture =
+      completionsDao.getAllCompletions(requestedRange.startExclusive, requestedRange.endInclusive)
+    val updateCacheRequest = allCompletionsFuture.map { fetchedCompletions =>
```

Comment:
Naming: `update` is a verb, but here we want an adjective desribed the request: `updatedCacheRequest` does that.